File: src/leetcode/MaxArea.jl
"""
    MaxArea(height::Vector{Int})::Int

Calculate the maximum area of water that can be contained between two lines.

# Arguments
- `height::Vector{Int}`: A vector of integers representing the height of lines.

# Returns
- `Int`: The maximum area of water that can be contained.

# Example
```julia
heights = [1, 8, 6, 2, 5, 4, 8, 3, 7]
println(MaxArea(heights))  # Output: 49
"""

function MaxArea(height::Vector{Int})::Int
    # Initialize left and right pointers to be at position 1 and n

    l, r = 1, length(height)
    max_area = 0
    while l < r
        max_area = max(max_area, min(height[l], height[r]) * (r - l))
        if height[l] < height[r]
            l += 1 # move to the center to find a higher height
        else
            r -= 1 # move to the center to find a higher height
        end
    end
    return max_area
end

File: src/leetcode/two_pointers/max_area.py
from typing import List


class Solution:
    def maxArea(self, height: List[int]) -> int:

        left, right = 0, len(height) - 1

        max_area = 0

        while left < right:
            current_area = (right - left) * min(height[left], height[right])
            max_area = max(current_area, max_area)

            if height[left] < height[right]:
                left += 1
            else:
                right -= 1

        return max_area


if __name__ == "__main__":
    sol = Solution()
    print(sol.maxArea([1, 8, 6, 2, 5, 4, 8, 3, 7]))
